### PR TITLE
[FreeBSD] Fix a few issues

### PR DESCRIPTION
**Pull request by antonmes**
This PR fixes a few issues on FreeBSD 11 / 12
1. creation of installable .tar.gz package
2. libswiftCore.so: undefined reference to
`__atomic_store` `__atomic_compare_exchange` `__atomic_load` *
3. crash of produced executables: `dlopen() failed on ./main: (NULL)Abort (core dumped)`

Related issues: https://github.com/apple/swift/pull/10836#issuecomment-317014783 

\* Couldn't find a better way.

### CODE REVIEW

**Comment by dcci** [comment]
@swift-ci please test

**Comment by dcci** [comment]
The FreeBSD bits look good to me but please wait for explicit approval by somebody on the swift team before merging.

**Comment by dcci** [comment]
@antonmes your code is failing on Linux, PTAL.

**Comment by antonmes** [comment]
@dcci I think it's not related to the PR, something with swift-corelibs-foundation in regression tests. Swift itself passes all the tests. This additions shouldn't affect any Darwin or Linux build instructions.

File: stdlib/public/runtime/ImageInspectionELF.cpp
Comment (dcci) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: Not extremely happy about this macro dance but it seems consistent with what's there (and probably OK).

File: cmake/modules/AddSwift.cmake
Comment (dcci) [inline cmake/modules/AddSwift.cmake]: Can you please add a comment here explaining why we need this?

**Review by dcci — CHANGES_REQUESTED** [review]

**Comment by dcci** [comment]
You probably need to wait until the failure goes away, rebase on top of master and retry. Some minors, otherwise I think this is fine.

File: cmake/modules/AddSwift.cmake
Comment (antonmes) [reply cmake/modules/AddSwift.cmake]: I thought the existing comment is enough, it's the same issue but we use builtins instead of `-latomic`. I don't know how to explain this or the reason. 😊

File: stdlib/public/runtime/ImageInspectionELF.cpp
Comment (gparker42) [reply stdlib/public/runtime/ImageInspectionELF.cpp]: Why is this necessary on FreeBSD? It is necessary on Android because the OS hides the contents of `/system/lib` from `dlopen()`.

File: cmake/modules/AddSwift.cmake
Comment (gparker42) [reply cmake/modules/AddSwift.cmake]: You should update the comment, then.

Comment (antonmes) [reply cmake/modules/AddSwift.cmake]: I mean it imports `__atomic_store` `__atomic_compare_exchange` `__atomic_load` from builtins. should i just list them there?

File: stdlib/public/runtime/ImageInspectionELF.cpp
Comment (antonmes) [reply stdlib/public/runtime/ImageInspectionELF.cpp]: I'm not sure about the reasons but this part is called and fails with `/libexec/ld-elf.so.1` and then with the main executable path. As for the latter, from https://github.com/apple/swift/pull/10836#issuecomment-317014783 I assumed that `dl_iterate_phdr` has the same behavior on FreeBSD.

**Comment by antonmes** [comment]
@swift-ci please smoke test

**Comment by dcci** [comment]
@swift-ci please test

File: cmake/modules/AddSwift.cmake
Comment (jrose-apple) [inline cmake/modules/AddSwift.cmake]: This seems like a very specific path to hardcode, but I'll take your word for it that it's a reasonable. (Especially for a CMake section we'd like to clean up later.)

**Comment by jrose-apple** [comment]
If @gparker42 has no further objections then I think this can go in. It's certainly not going to break anyone else.

File: stdlib/public/runtime/ImageInspectionELF.cpp
Comment (johnno1962) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: Any chance you could change this to “/system” while you’re in the neighbourhood as there are libraries that are blocked on some devices in “/system/vendor” as well as “/system/lib". Non of these libraries will have Swift implementations to worry about. Saves me a PR.

Comment (jrose-apple) [reply stdlib/public/runtime/ImageInspectionELF.cpp]: :-/ Let's not mix FreeBSD and Android changes.

Comment (johnno1962) [reply stdlib/public/runtime/ImageInspectionELF.cpp]: :-/

**Comment by gparker42** [comment]
Looks good. @dcci?

**Comment by dcci** [comment]
LGTM, thanks!

**Review by dcci — APPROVED** [review]

File: cmake/modules/AddSwift.cmake
Comment (dpgao) [reply cmake/modules/AddSwift.cmake]: This hard-coded path does not work on my build. The `.a` file has been moved to `build/Ninja-ReleaseAssert/llvm-freebsd-x86_64/lib/clang/5.0.0/lib/freebsd/libclang_rt.builtins-${arch}.a`.

Comment (jrose-apple) [reply cmake/modules/AddSwift.cmake]: Ah, `${SWIFTLIB_DIR}/clang` should be a symlink to `lib/clang/5.0.0`, so if you're seeing issues here, something else has gone wrong.

Comment (jrose-apple) [reply cmake/modules/AddSwift.cmake]: Note that if you haven't built in a long time, the Clang version of the symlink may not match the actual Clang version.

Comment (jrose-apple) [reply cmake/modules/AddSwift.cmake]: It's possible you just weren't using anything (from C) that needed atomics.

**Pull request merged**
